lib/controller/controller.go

```diff
@@ -716,75 +698,6 @@ func (ctrl *ProvisionController) shouldDelete(volume *v1.PersistentVolume) bool
 	return true
 }
 
-// lockProvisionClaimOperation wraps provisionClaimOperation. In case other
```

Comment:
The issue if we remove it without a replacement is we will be trading kube API abuse for storage backend API abuse, in case somebody is running more than one provisioner instance (which is way too easy given they can just e.g. change deployment # from 1 to 2). I.e. instead of racing to lock a PVC and spamming the kube API server they will race to talk to the storage backend then race to create a PV, so we need a replacement per-storageclass leader election if we want to remove this.

Also if we get rid of this leader election there is a lot more code that can be removed, which I will be happy to be rid of, but I will do it myself in a subsequent and I would appreciate a review!

Reply:
1. will be much more difficult since there are many other provisioner besides csi external-provisioner depending on external-storage and they will each be burdened with copying the same complementation, even if the implementation is simple. 2 is better I think. The more opaque all of this is to library consumers the better. I don't like the idea of our little controller depending on some configmap for maintaining leader state but I also don't want to overload storage classes for that purpose like we are overloading pvcs at the moment. I will have more time to think/work on this in the coming days

Reply:
Agree with you. From the implementation perspective, `external-storage` could copy kind of leader election as existing [`operator`](https://github.com/coreos/etcd-operator/blob/c8f63d508266990a4d20718d94363c30e75e6282/cmd/operator/main.go#L121) does. 

Any new ideas, please let me know. Many thanks

Reply:
hi, in which scenario current `per-PVC` lock is bad? I like `per-PVC` lock idea, because I can simply deploy more provisioners to scale. Each provisioner can work independently.

Reply:
Each provisioner tries to write a PVC's annotation in order to be a leader. If failed, each one will retry after some time and lead to a race condition. As a matter of fact, the implementation of `per-PVC` lock hinders `external-provisioner` scalable.

Reply:
Thank you @cofyc for chiming in, i was truly ready to dump per-pvc lock assuming nobody uses the feature. :) So we don’t want per-pvc lock because it scales poorly wrt apiserver but we don’t want to switch to per-cluster or per-class because that prevents simple horizontal scaling of provisioners. 

I am still leaning toward increasing the lock granularity. Horizontal scaling is nice to have but in practice I have to wonder how much it helps. Controller-manager is just one instance but it doesn’t choke with many pvcs, why should our little controller be any different.

In the beginning sharding was seen as the ideal solution but I could not find a feasible way to do it. If we are going to be writing to an external endpoint/configmap anyway, perhaps there is something fancier than plain leader election that could enable us to do sharding but I think that is a dangerous idea :)

Reply:
It would also help me to understand how apiserver throttling happens. Is it the burst of x ‘Updates’ at once that kills it? Would jitter or back off help alleviate the problem? I’ll research this tomorrow....

Reply:
From my point of view, `active-passive` could be more proper in practice than `sharding`.

Reply:
I am not convinced client-side throttling is even a problem anymore after we introduced PVC work queues, since `Provision` typically takes a lot of time. But I am never opposed to adding more construct functions. 

I've opened https://github.com/kubernetes-incubator/external-storage/issues/847 to discuss operator principles in general.

I think we need to bring this up with more people, e.g. in the next sig-storage meeting if there's time, otherwise I will ramble on indecisively forever. We should have this resolved AT LATEST before 1.12 release IMO

Reply:
I'll also write an email to sig-storage google group next week

Reply:
@wongma7 any chance you can join us to discuss this issue ?

Reply:
@vladimirvivien yes, where?

Reply:
@wongma7 
if you can, join the next CSI meeting (Wednesday 10am PST) for a quick report on the stuff you are working (LE/informers).  Here is the link below:

VC on Zoom: https://zoom.us/j/614261834

Notes and agenda doc: https://docs.google.com/document/d/1-WmRYvqw1FREcD1jmZAOjC0jX6Gop8FMOzsdqXHFoT4/edit?usp=sharing